src/platforms/macos/daemon/macosdnsmanager.cpp

```diff
@@ -0,0 +1,266 @@
+    CFRelease(m_scStore);
+  }
+  MZ_COUNT_DTOR(MacOSDnsManager);
+}
+
+static QString cfParseString(CFTypeRef ref) {
+  if (CFGetTypeID(ref) != CFStringGetTypeID()) {
+    return QString();
+  }
+
+  CFStringRef stringref = (CFStringRef)ref;
+  CFRange range;
+  range.location = 0;
+  range.length = CFStringGetLength(stringref);
+  if (range.length <= 0) {
+    return QString();
+  }
+
+  UniChar* buf = (UniChar*)malloc(range.length * sizeof(UniChar));
+  if (!buf) {
+    return QString();
+  }
+  auto guard = qScopeGuard([&] { free(buf); });
+
+  CFStringGetCharacters(stringref, range, buf);
+  return QString::fromUtf16(buf, range.length);
+}
+
+static void cfDictSetString(CFMutableDictionaryRef dict, CFStringRef name,
+                            const QString& value) {
+  if (value.isNull()) {
+    return;
+  }
+  CFStringRef cfValue = CFStringCreateWithCString(
+      kCFAllocatorSystemDefault, qUtf8Printable(value), kCFStringEncodingUTF8);
+  CFDictionarySetValue(dict, name, cfValue);
+  CFRelease(cfValue);
+}
+
+static void cfDictSetStringList(CFMutableDictionaryRef dict, CFStringRef name,
+                                const QStringList& valueList) {
+  if (valueList.isEmpty()) {
+    return;
+  }
+
+  CFMutableArrayRef array;
+  array = CFArrayCreateMutable(kCFAllocatorSystemDefault, 0,
+                               &kCFTypeArrayCallBacks);
+  if (array == nullptr) {
+    return;
+  }
+
+  for (const QString& rstring : valueList) {
+    CFStringRef cfAddr = CFStringCreateWithCString(kCFAllocatorSystemDefault,
+                                                   qUtf8Printable(rstring),
+                                                   kCFStringEncodingUTF8);
+    CFArrayAppendValue(array, cfAddr);
+    CFRelease(cfAddr);
+  }
+  CFDictionarySetValue(dict, name, array);
+  CFRelease(array);
+}
+
+int MacOSDnsManager::run(QStringList& tokens) {
+  Q_ASSERT(!tokens.isEmpty());
+  QString appName = tokens[0];
+
+  QCoreApplication app(CommandLineParser::argc(), CommandLineParser::argv());
+
+  QCoreApplication::setApplicationName("Mozilla VPN DNS Manager");
+  QCoreApplication::setApplicationVersion(Constants::versionString());
+  QTextStream stream(stderr);
+
+  if (tokens.length() <= 1) {
+    stream << "Error: At least one DNS server address is required." << Qt::endl;
+    stream << Qt::endl;
+    stream << "Usage: [options] ADDR [ADDR ...]" << Qt::endl;
+    return 1;
+  }
+
+  // Ensure that the DNS addresses we received are valid.
+  QStringList dnsAddrList = tokens.mid(1);
+  for (const QString& addr : dnsAddrList) {
+    QHostAddress host = QHostAddress(addr);
+    if (host.isNull()) {
+      stream << "Error: DNS Server \'" << addr << "\' is invalid." << Qt::endl;
+      return 1;
+    }
+    if (host.isBroadcast() || host.isMulticast()) {
+      stream << "Error: DNS Server \'" << addr << "\' is a multicast address."
+             << Qt::endl;
+      return 1;
+    }
+    // TODO: Should we prohibit the loopback address?
```

Comment:
For my understanding, why would be the pros and cons of prohibiting that?

Reply:
Allowing `localhost` allows users to run a DNS server or proxy on their local machine to intercept and manipulate DNS (eg: something like a Pi-Hole), effectively rolling their own solution for DNS-based ad or tracker blocking. However, we have seen some platforms (eg: iOS and Android) explicitly prohibit this (see: `DNSHelper::validateUserDNS()`).

However, given that this is a MacOS-specific tool, perhaps it is fine to permit the loopback address after all.